File: papers/aplas/benchmarks/lTypechecker/original.pl
type_(bConst_(X), Gamma, some(boolean)).
type_(iConst_(Y), Gamma, some(integer)).
type_(var_(V), Gamma, Ttype) :- idx(V, Gamma, Ttype).
type_(plus_(M, N), Gamma, some(integer)) :- type_(M, Gamma, some(integer)), type_(N, Gamma, some(integer)).
type_(mult_(M, N), Gamma, some(integer)) :- type_(M, Gamma, some(integer)), type_(N, Gamma, some(integer)).
type_(equal_(L, R), Gamma, some(boolean)) :- type_(L, Gamma, some(T1)), type_(R, Gamma, some(T1)).
type_(less_(L, R), Gamma, some(boolean)) :- type_(L, Gamma, some(integer)), type_(R, Gamma, some(integer)).
type_(let_(Bound, Body), Gamma, Ttype) :- type_(Bound, Gamma, some(Btype1)), type_(Body, cons(Btype1, Gamma), Ttype).
type_(if_(Cond, Thn, Els), Gamma, Ttype) :- type_(Cond, Gamma, some(boolean)), type_(Thn, Gamma, Ttype), type_(Els, Gamma, Ttype).
idx(o, nil, none).
idx(o, cons(H, T), some(H)).
idx(s(K1), cons(H, T), V) :- idx(K1, T, V).

File: papers/aplas/benchmarks/lTypechecker/original.ml
open GT
open OCanren
open OCanren.Std
open Helper

let topLevel x0 = let rec type_ term gamma ttype = (fresh (els thn cond btype1 btype body bound t1 t r l n m v y x) ((((term === bConst_ x) &&& (ttype === (Option.some (boolean ())))) ||| (((term === iConst_ y) &&& (ttype === (Option.some (integer ())))) ||| (((term === var_ v) &&& (idx v gamma ttype)) ||| (((term === plus_ m n) &&& ((type_ m gamma ((Option.some (integer ())))) &&& ((type_ n gamma ((Option.some (integer ())))) &&& (ttype === (Option.some (integer ())))))) ||| (((term === mult_ m n) &&& ((type_ m gamma ((Option.some (integer ())))) &&& ((type_ n gamma ((Option.some (integer ())))) &&& (ttype === (Option.some (integer ())))))) ||| (((term === equal_ l r) &&& ((type_ l gamma t) &&& ((type_ r gamma t) &&& ((ttype === (Option.some (boolean ()))) &&& (t === (Option.some (t1))))))) ||| (((term === less_ l r) &&& ((type_ l gamma ((Option.some (integer ())))) &&& ((type_ r gamma ((Option.some (integer ())))) &&& (ttype === (Option.some (boolean ())))))) ||| (((term === let_ bound body) &&& ((type_ bound gamma btype) &&& ((btype === (Option.some (btype1))) &&& (type_ body ((Std.(%) (btype1) (gamma))) ttype)))) ||| ((term === if_ cond thn els) &&& ((type_ cond gamma ((Option.some (boolean ())))) &&& ((type_ thn gamma ttype) &&& (type_ els gamma ttype)))))))))))))) and idx k xs v = (fresh (k1 t h) ((((k === Std.Nat.zero) &&& ((xs === (Std.List.nil ())) &&& (v === (Option.none ())))) ||| (((k === Std.Nat.zero) &&& ((xs === (Std.(%) (h) (t))) &&& (v === (Option.some (h))))) ||| ((k === (Std.Nat.succ (k1))) &&& ((xs === (Std.(%) (h) (t))) &&& (idx k1 t v))))))) in   (((type_ x0 ((Std.List.nil ())) ((Option.some (integer ()))))))
